### ISSUE TITLE
Trend charts: x-axis labelling issue

### ISSUE THREAD

**Issue by mmontesanonyc** [nychealth/EH-dataportal#1080]
We have some difficulty with trend-chart x-axis labelling. 

**We need to figure out a way to handle both x-axis consistency/inconsistency, and x-axis as string and numeric values.** For data vis/comms standards, we obviously want even time intervals represented by even dimension intervals. But currently, the VL spec treats the x-axis as a string, not a numeric value, so it places each x-axis interval equidistant from each other.

For this AQ indicator, we see string labels for the x-axis - 'annual average 2015' et cetera. 
<img width="866" alt="image" src="https://github.com/user-attachments/assets/d4046214-2cf9-43cb-b524-d1f58f8d9ea3">

For this asthma indicator, we have null data for 2015 to affect the gap, and despite inconsistent data we have consistent x-axis intervals.
<img width="811" alt="image" src="https://github.com/user-attachments/assets/76da2c02-fde1-4d6e-9a5c-2ff0c647e737">

But for this AC (CHS) indicator, we see the issue profoundly - with uneven intervals of available data, and without null data, we have a wonky trend chart where the distance from 2007 to 2020 is the same as the distance from 2020 to 2021. If the VL spec were reading the x-axis labels as numerals instead of strings, then we wouldn't have this issue - but it treats them as strings.
<img width="866" alt="image" src="https://github.com/user-attachments/assets/3a3c32d0-7b86-428d-8ec9-649ee7241103">

**Comment by cgettings** [comment]
**TL;DR:** There's always a chance that some Vega-Lite way exists to do this, like maybe some layer or transform magic, but I haven't found it yet. I'm ok getting weird if it means not having to reconstruct from scratch a value we already have in the same data object.
_________________
The main challenge here is that Vega-Lite doesn't allow you to use different fields for the value vs. label of the encoding channel: The label has to be either the value of the `field` property of the encoding channel, or some derivation of it. So for the x-axis, we can't give `end_period` as the value, and `TimePeriod` as the label. The label has to be derived by parsing `end_period` as a datetime and then reproducing `TimePeriod` using datetime formatting. Given that `TimePeriod` is bespoke, this is at best annoying, and at worst impossible.

If it's not impossible, we could figure out some way of parsing `TimePeriod` and using a conditional to inline a corresponding format in the spec. Annoying!

**Comment by mmontesanonyc** [comment]
Wonder if we can think through a way where we just use `end_period` and keep `TimePeriod` in the tooltip; this might be slightly undesirable for PM2.5 and for ACS indicators (multi-year data), but that might be less bad than how this handles CHS inconsistency.

**Comment by cgettings** [comment]
A potential layer solution is adding a text mark at `(end_period, 0)` (maybe with a negative y offset) whose text is `TimePeriod`. We could probably make it look identical to proper tick labels, and it probably even works the same under the hood, which would make it harder to break.

**Comment by cgettings** [comment]
[Gist link](https://gist.github.com/cgettings/4e70cc8250ede6bcfb9af80635dcb0a3)
[Vega editor link](https://vega.github.io/editor/?#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoMcGmSxoxABk1KA7jQAm9NAA5tSg0kwpUoYkkEM4ENAG1QcWQYD6OOACd2AzQARjEAFgA2ACYQ8PDtRKUAFRoEOAAFAKC0EABBRRAANQcnNE0AXwUPL19stmDUMIB2Vq0ExJ0QVPSswIbcgCFCksc4NGaqmp8-fsaQyM0ATgBmZuizTpS0zPrGkABhEdLx1DEpkE8ZvdDIhKXNE02k7p2+nPQAEWOx0JCLq51Oa3dbxcLPcyvXo3dAAUR+ZVQ4QqAF0qiBBEgAJ4BNygZD+ADWuWEsnGSk8UAaNFkZDQoAAHvSQAAzGhwQT7QGzHJKTBYvy5TCIHBsfwOQqYHYAVVkdFyOKQ-kldEEp1kDEEgiUSAZNBctgxSCYHINLIcEDgSjIgUa5sElr5NCghINmH8Tgq6KxzLZHP2ozKfIFpxAAEcGEhZFLrFLSCAvRcCcT0MKGZhCpTqbTmUzDX7ObluTdg4LUyKxRKnelZfL0IrlU7MGq0BqtTq9QbQJiTQ60PbHSAbYZ+xarSApS63R64F6lD7DfZfmdopU+XB0772YX0D1dsDS6HZGwEDSJYm0UA)

```
{
    "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
    "height": 500,
    "width": 800,
    "data": {
        "values": [
            {"end_period": 1546214400000, "TimePeriod": "A", "Value": 0},
            {"end_period": 1577750400000, "TimePeriod": "B", "Value": 7},
            {"end_period": 1609372800000, "TimePeriod": "C", "Value": 5},
            {"end_period": 1640908800000, "TimePeriod": "D", "Value": 11},
            {"end_period": 1672444800000, "TimePeriod": "E", "Value": 4}
        ]
    },
    "layer": [
        {
            "mark": "line",
            "encoding": {
                "x": {
                    "field": "end_period",
                    "type": "quantitative",
                    "title": null,
                    "axis": {"labels": false, "grid": false, "ticks": false}
                },
                "y": {"field": "Value", "type": "quantitative"}
            }
        },
        {
            "mark": "text",
            "encoding": {
                "x": {
                    "field": "end_period",
                    "type": "quantitative",
                    "axis": {"labels": false, "grid": false, "ticks": false}
                },
                "y": {"value": 520},
                "text": {"field": "TimePeriod", "type": "nominal"}
            }
        },
        {
            "mark": "tick",
            "encoding": {
                "x": {
                    "field": "end_period",
                    "type": "quantitative",
                    "axis": {"labels": false, "grid": false, "ticks": true}
                },
                "y": {"value": 500},
                "color": {"value": "black"}
            }
        },
    ]
}
```

**Comment by cgettings** [comment]
Label formatting looks weird on the site, but other than that... success?

![visualization](https://github.com/user-attachments/assets/b42e1e0e-bc7c-4c0e-83e4-bba0fd8d11ce)

**Comment by mmontesanonyc** [comment]
Amazing! In the past I've had some funky experiences rendering text marks - difficulty styling - but on my browser these are looking great.

Here are the 4 conditions I'm testing
- [PM2.5, annual averages](/data-explorer/air-quality/?id=2023#display=trend) - long strings
- [Asthma ed visits, adults](/data-explorer/asthma/?id=2380#display=trend) - discontinuity plus null
- [Adults with AC](/data-explorer/housing-safety/?id=2104#display=trend) - discontinuity, no null
- [Neighborhood poverty](/data-explorer/economic-conditions/?id=103#display=trend) - multi-year

The long strings condition is problematic. But I wonder if there's an opportunity to chat with AQ folks about re-naming that - if the measure is "Annual average mean" then we don't need "Annual average" repeated on each tick. 

Another thing we might do is label every other tick (there may be a necessary mobile fallback, too). Or, we can do a calculate transform that [splits the string](https://stackoverflow.com/questions/64766369/vega-lite-line-break-or-word-wrap-for-labels-in-axis) - [more here](https://gramener.medium.com/text-wrapping-in-vega-js-charts-svg-and-canvas-197703b443da).

**Comment by mmontesanonyc** [comment]
Added a line that handles wrapping - I think we're getting there!

<img width="857" alt="image" src="https://github.com/user-attachments/assets/617ff915-2589-439a-bdb0-c034bde2c18a">

**Comment by mmontesanonyc** [comment]
One option for mobile fallback - this is another calculate transform that just grabs the year (specifically, the last item in the array) and, on small screen sizes, uses that as the x-axis label. We still use `TimePeriod` in the tooltip. 

Still, ACS indicators or anything else with multi-year labels (eg `2017-2019`) are imperfect on mobile.

**Comment by mmontesanonyc** [comment]
Perhaps a better option for mobile fallback is derive the Year from `end_period`, and another field that runs a test - if Year is divisible by 2, then its value is `TimePeriod` and if not, it's ' '. That will effectively label every other interval.

**Comment by cgettings** [comment]
The "no_compare" line is _on_ the year, instead of before it:

![image](https://github.com/user-attachments/assets/0a8028ef-2aff-4a12-8d18-1453c46136b5)

**Comment by mmontesanonyc** [comment]
ah, good catch. that should be handle-able.

**Comment by mmontesanonyc** [comment]
This has largely been folded into work on #1127, so I'm going to close this as broader work has continued under that issue.
